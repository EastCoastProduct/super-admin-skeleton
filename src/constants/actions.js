export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const LOGOUT_SUCCESS = 'LOGOUT_SUCCESS';

export const USERS_GET_FAILED = 'USERS_GET_FAILED';
export const USERS_GET_SUCCESS = 'USERS_GET_SUCCESS';
export const USERS_PAGINATION_CHANGE = 'USERS_PAGINATION_CHANGE';

export const USER_GET_FAILED = 'USER_GET_FAILED';
export const USER_FETCH_SUCCESS = 'USER_FETCH_SUCCESS';
