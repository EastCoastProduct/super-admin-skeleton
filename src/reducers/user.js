import { fromJS } from 'immutable';
import { USER_FETCH_SUCCESS, USER_GET_FAILED, USER_CHANGE_SUCCESS }
  from '../constants/actions';

const initialState = fromJS({
  error: null,
  profile: {},
});

export default (state = initialState, action) => {
  switch (action.type) {
    case USER_FETCH_SUCCESS:
      return state.merge(fromJS({
        error: null,
        profile: action.profile,
      }));
    case USER_CHANGE_SUCCESS:
      return state.merge(fromJS({
        error: null,
        profile: state.get('profile').merge(...action.responses),
      }));
    case USER_GET_FAILED:
      return state.merge(fromJS({
        error: action.error,
      }));
    default:
      return state;
  }
};
