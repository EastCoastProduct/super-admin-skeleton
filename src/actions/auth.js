import store from 'store';
import { LOGIN_SUCCESS, LOGOUT_SUCCESS } from '../constants/actions';
import { API_URL } from '../constants/application';
import parseErrors from '../utils/parseErrors';
import fetch from '../utils/fetch';

export const loginSuccess = superadmin => ({
  type: LOGIN_SUCCESS,
  superadmin,
});

export const logoutSuccess = () => ({
  type: LOGOUT_SUCCESS,
});

export const loginFetch = (values, cb) =>
  dispatch =>
    fetch(`${API_URL}/superAdmin/authenticate`, {
      method: 'POST',
      body: JSON.stringify(values),
    }).then((resp) => {
      store.set('token', `Bearer ${resp.token}`);
      store.set('superadmin', resp.user);
      dispatch(loginSuccess(resp.user));
      return typeof cb === 'function' && cb();
    }).catch(err =>
      Promise.reject(parseErrors(err))
    );

export const logoutAction = cb =>
  (dispatch) => {
    store.clear();
    dispatch(logoutSuccess());
    return typeof cb === 'function' && cb();
  };
