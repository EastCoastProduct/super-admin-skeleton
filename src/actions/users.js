import queryString from 'query-string';
import { USERS_GET_SUCCESS, USERS_GET_FAILED, USERS_PAGINATION_CHANGE }
  from '../constants/actions';
import { API_URL } from '../constants/application';
import fetch from '../utils/fetch';

export const usersGetSuccess = (list, listTotal) => ({
  type: USERS_GET_SUCCESS,
  list,
  listTotal,
});

export const usersGetFailed = error => ({
  type: USERS_GET_FAILED,
  error,
});

export const paginationChange = page => ({
  type: USERS_PAGINATION_CHANGE,
  page,
});

export const usersGetFetch = (qs, cb) =>
  dispatch =>
    fetch(`${API_URL}/superAdmin/users?${queryString.stringify(qs)}`)
      .then((resp) => {
        dispatch(usersGetSuccess(resp.rows, resp.count));
        return typeof cb === 'function' && cb();
      }).catch(err =>
        dispatch(usersGetFailed(err.message))
      );
