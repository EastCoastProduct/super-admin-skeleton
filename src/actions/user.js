import { USER_FETCH_SUCCESS, USER_GET_FAILED } from '../constants/actions';
import { API_URL } from '../constants/application';
import createFormData from '../utils/createFormData';
import parseErrors from '../utils/parseErrors';
import fetch from '../utils/fetch';

export const userFetchSuccess = profile => ({
  type: USER_FETCH_SUCCESS,
  profile,
});

export const userGetFailed = error => ({
  type: USER_GET_FAILED,
  error,
});

export const userGetFetch = userId =>
  dispatch =>
    fetch(`${API_URL}/users/${userId}`).then(resp =>
      dispatch(userFetchSuccess(resp))
    ).catch(err =>
      dispatch(userGetFailed(err.message))
    );

export const userUpdateFetch = (values, userId, cb) =>
  dispatch =>
    fetch(`${API_URL}/users/${userId}`, {
      method: 'POST',
      body: createFormData(values),
    }).then((resp) => {
      dispatch(userFetchSuccess(resp));
      return typeof cb === 'function' && cb();
    }).catch(err =>
      Promise.reject(parseErrors(err))
    );

export const userCreateFetch = (values, cb) =>
  dispatch =>
    fetch(`${API_URL}/superAdmin/users`, {
      method: 'POST',
      body: JSON.stringify(values),
    }).then((resp) => {
      dispatch(userFetchSuccess(resp));
      return typeof cb === 'function' && cb(resp.id);
    }).catch(err =>
      Promise.reject(parseErrors(err))
    );
