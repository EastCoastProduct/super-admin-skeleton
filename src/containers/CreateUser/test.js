import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';
import deepMerge from 'deepmerge';
import { CreateUserComponent } from './';
import * as Actions from '../../actions/user';

describe('CreateUser component', () => {
  const id = 1;
  const props = {
    dispatch: jest.fn(() => Promise.resolve(id)),
    form: 'CreateUser',
    handleSubmit: f => f,
    router: {
      push: jest.fn(),
    },
    submitting: true,
  };
  const wrapper = shallow(<CreateUserComponent {...props} />);

  describe('snapshot', () => {
    it('renders required data', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders required data with error', () => {
      const newProps = deepMerge(props, {
        error: 'User already exists.',
        submitting: false,
      });
      const newWrapper = shallow(<CreateUserComponent {...newProps} />);

      expect(newWrapper).toMatchSnapshot();
    });
  });

  describe('instance', () => {
    afterEach(() => {
      jest.resetAllMocks(); // move to before when clearAllMocks gets out
    });
    Actions.userCreateFetch = jest.fn();
    const instance = wrapper.instance();

    it('handleUserCreate method', () => {
      const values = fromJS({
        email: 'test@email.com',
      });
      const p = instance.handleUserCreate(values);

      expect(Actions.userCreateFetch).toHaveBeenCalledWith(values);
      expect(instance.props.dispatch).toHaveBeenCalled();
      return p.then(() => {
        expect(instance.props.router.push).toHaveBeenCalledWith(`/user/${id}`);
      });
    });
  });
});
