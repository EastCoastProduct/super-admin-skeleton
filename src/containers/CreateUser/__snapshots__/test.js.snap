exports[`CreateUser component snapshot renders required data 1`] = `
<UserBox
  header="Create New User">
  <form
    noValidate={true}
    onSubmit={[Function]}>
    <Field
      component={[Function]}
      id="CreateUser"
      label="Email"
      name="email"
      placeholder="Email"
      type="email"
      validate={
        Array [
          [Function],
          [Function],
        ]
      }
      validated={true} />
    <BoxButtons
      disabled={true}
      link="/">
      Create
    </BoxButtons>
  </form>
</UserBox>
`;

exports[`CreateUser component snapshot renders required data with error 1`] = `
<UserBox
  header="Create New User">
  <form
    noValidate={true}
    onSubmit={[Function]}>
    <Field
      component={[Function]}
      id="CreateUser"
      label="Email"
      name="email"
      placeholder="Email"
      type="email"
      validate={
        Array [
          [Function],
          [Function],
        ]
      }
      validated={true} />
    <BoxButtons
      disabled={false}
      link="/">
      Create
    </BoxButtons>
    <ErrorMsg>
      User already exists.
    </ErrorMsg>
  </form>
</UserBox>
`;
